### PR TITLE
POWER10: Optimized SGEMM in MLAS

### PR DESCRIPTION
**Pull request by RajalakshmiSR**
This patch introduces new optimized version of SGEMM in MLAS
using power10 Matrix-Multiply Assist (MMA) feature introduced in
POWER ISA v3.1. This patch makes use of new POWER10 compute instructions
for matrix multiplication operation.

### CODE REVIEW

**Comment by RajalakshmiSR** [comment]
@tracysh Requesting review.

File: cmake/onnxruntime_mlas.cmake [hunk 1]
```diff
@@ -202,6 +202,29 @@ else()
     set(mlas_platform_srcs
       ${ONNXRUNTIME_ROOT}/core/mlas/lib/power/SgemmKernelPower.cpp
     )
+    check_cxx_compiler_flag("-mcpu=power10" HAS_POWER10)
+    if(HAS_POWER10)
+      set(CMAKE_REQUIRED_FLAGS "-mcpu=power10")
+      check_cxx_source_compiles("
+        #include <altivec.h>
+        int main() {
+	  __vector_quad acc0;
```
Comment (yufenglee) [inline cmake/onnxruntime_mlas.cmake:211]: >	 [](http://example.com/codeflow?start=0&length=1)

nit: space instead of tab

Comment (yufenglee) [reply cmake/onnxruntime_mlas.cmake:211]: and tabs below

Comment (RajalakshmiSR) [reply cmake/onnxruntime_mlas.cmake]: @yufenglee Thanks. Fixed it.

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+
+        const float* a = A;
+        size_t k = CountK;
+
+        MLAS_FLOAT32X4 Accumulators[2][RowCount] = {{ 0 }};
+        MLAS_FLOAT32X4 Result[RowCount];
+        MLAS_FLOAT32X4 AElements[RowCount];
+        MLAS_FLOAT32X4 ABroadcast[RowCount] = { 0 };
+        MLAS_FLOAT32X4 A2Broadcast[RowCount] = { 0 };
+	__vector_quad acc[8];
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:567]: >	 [](http://example.com/codeflow?start=0&length=1)

nit: space and indent

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+{
+	__vector float a1,a2;
+	a1 = vec_mergee (AElements[0], AElements[1]);
+	a2 = vec_mergee (AElements[2], AElements[3]);
+	ABroadcast[0] =vec_xxpermdi(a1,a2,0);
+	ABroadcast[2] =vec_xxpermdi(a1,a2,3);
+	a1 = vec_mergeo (AElements[0], AElements[1]);
+	a2 = vec_mergeo (AElements[2], AElements[3]);
+	ABroadcast[1] =vec_xxpermdi(a1,a2,0);
+	ABroadcast[3] =vec_xxpermdi(a1,a2,3);
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:206]: nit: space

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+};
+template<size_t VectorCount>
+struct MlasSgemmStoreVectorMMA
+{
+    template<size_t RowCount, size_t Row>
+    MLAS_FORCEINLINE
+    static
+    void
+    Iteration(
+	MLAS_FLOAT32X4 Result[4],
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:291]: >	 [](http://example.com/codeflow?start=0&length=1)

nit: similar here

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp (hunk 2 above)
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:567]: >__vector_quad [](http://example.com/codeflow?start=1&length=13)

what's the size of __vector_quad? 4*128 bits? I don't find a wiki on it

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+    BElements[3] = MlasLoadFloat32x4(B + 12);
+
+    MlasLoopUnroll<RowCount, MlasSgemmMultiplyAddRow>()(Accumulators, ABroadcast, BElements);
+}
+template<size_t RowCount>
+MLAS_FORCEINLINE
+void
+MlasSgemmComputeAElements(
+    MLAS_FLOAT32X4 AElements[RowCount],
+    MLAS_FLOAT32X4 ABroadcast[RowCount]
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:195]: >RowCount [](http://example.com/codeflow?start=30&length=8)

are you assuming RowCount is 4 here? If I understand correctly, AElements is transposed to ABroadcast

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp (hunk 2 above)
Comment (RajalakshmiSR) [reply onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:567]: __vector_quad is a 64-byte accumulator and introduced in gcc  here 
https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=f002c046e37d0027513af5297d9259e1fad29c27

Also for builtins, https://gcc.gnu.org/onlinedocs/gcc/PowerPC-Matrix-Multiply-Assist-Built-in-Functions.html

File: onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp (hunk 5 above)
Comment (RajalakshmiSR) [reply onnxruntime/core/mlas/lib/power/SgemmKernelPOWER10.cpp:195]: Yes RowCount is always 4 for MlasSgemmComputeAElements.

**Comment by yufenglee** [comment]
It looks good to me. How much performance speed up do you see with MMA?

**Comment by RajalakshmiSR** [comment]
@yufenglee Thanks for the review. The preliminary standalone MLAS tests show more than 50% improvements.

**Review by yufenglee — APPROVED** [review]

**Comment by yufenglee** [comment]
@RajalakshmiSR, I synced up with @tracysh offline and he reminded of that mlas/lib/power/SgemmKernelPower.cpp already has an implementation for MlasSgemmProcessCount, which can be resued.

**Comment by RajalakshmiSR** [comment]
@tracysh @yufenglee I added MlasSgemmMMAProcessCount for sizes 4 and 8 and using same MlasSgemmProcessCount for sizes 2 & 1. Do you mean to avoid duplicate MlasSgemmProcessCount () in SgemmKernelPOWER10.cpp ?

**Comment by yufenglee** [comment]
> @tracysh @yufenglee I added MlasSgemmMMAProcessCount for sizes 4 and 8 and using same MlasSgemmProcessCount for sizes 2 & 1. Do you mean to avoid duplicate MlasSgemmProcessCount () in SgemmKernelPOWER10.cpp ?

yes, and other duplications like MlasLoopUnrollStep.

**Comment by RajalakshmiSR** [comment]
@yufenglee @tracysh Added new commit based on the last comment.

**Review by yufenglee — APPROVED** [review]

**Comment by yufenglee** [comment]
Thanks @RajalakshmiSR!

**Comment by RajalakshmiSR** [comment]
@yufenglee @snnn Will this fix be available in 1.8.2?  Is there anything that I have to do to make this commit available in the next release?

**Comment by snnn** [comment]
Sorry currently we don't have a plan for 1.8.2.  It may not happen.

**Comment by RajalakshmiSR** [comment]
@snnn Then I guess it should be in the 1.9.0 release. Do you have any idea about the 1.9.0 release date?

**Comment by snnn** [comment]
https://github.com/microsoft/onnxruntime/projects/5 . It would be in this fall.

**Pull request merged**